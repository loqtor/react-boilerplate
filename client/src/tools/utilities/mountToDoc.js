/*
 * Utility to provide aXe with a fake DOM to allow for accessibility testing
 * on mounted components within our Jest tests
*/
import { mountWithIntl } from './enzymeIntlTestHelper';

let wrapper;

export default function mountToDoc(component) {
  const container = mountWithIntl(component);

  if (!wrapper) {
    wrapper = document.createElement('main');
    document.body.appendChild(wrapper);
  }

  wrapper.innerHTML = '';
  wrapper.appendChild(container.getDOMNode());

  return container;
}
