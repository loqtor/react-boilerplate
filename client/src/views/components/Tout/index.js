import React, { Component } from 'react';
import classnames from 'classnames';
import onClickOutside from 'react-onclickoutside';
import PropTypes from 'prop-types';

import { Button } from '../Button';

class Tout extends Component {
  state = {
    isOpen: false,
  };

  toggleTout = () => {
    this.setState(prevState => ({ isOpen: !prevState.isOpen }));
  };

  handleClickOutside = () => {
    const { isOpen } = this.state;

    if (isOpen) {
      this.setState({ isOpen: false });
    }
  };

  render() {
    const { isOpen } = this.state;
    const {
      additionalClasses,
      alertType,
      children,
      icon,
      id,
      isIconOnly,
      isInline,
      isToolTip,
      label,
      labelClassName,
      onClick,
      popupClassName,
      position,
      size,
    } = this.props;
    const isBottom = ['bottomLeft', 'bottomCenter', 'bottomRight'].includes(position);
    const popupName = classnames(popupClassName, `Tout-popup is-${position}`);
    const toutClassName = classnames('Tout', additionalClasses, {
      'is-open': isOpen,
      'Tout--alert': alertType === 'alert',
      'Tout--confirmation': alertType === 'confirmation',
      'Tout--error': alertType === 'error',
      'Tout--info': alertType === 'info',
      'Tout--inline': isInline,
      'Tout--toolTip': isToolTip,
    });
    const promptClassName = classnames('Tout-prompt', labelClassName, {
      'Tout-prompt--textSmall': size === 'small',
      'is-bottom': isBottom,
    });

    if (isInline) {
      return (
        <div className={toutClassName} role="tooltip" id={id}>
          {alertType && <span className={`icon icon-${alertType}`} />}
          {children}
        </div>
      );
    }

    if (isToolTip) {
      return (
        <div className={toutClassName} id={id}>
          {children || (
            <Button
              icon={icon}
              isIconOnly={isIconOnly}
              className={promptClassName}
              buttonType="button"
              label={label}
              onClick={onClick || (() => this.toggleTout())}
              onBlur={onClick ? null : () => this.toggleTout()}
            />
          )}
          <div role="tooltip" aria-hidden="true" className={popupName}>
            {label}
          </div>
        </div>
      );
    }

    return (
      <div className={`${toutClassName} ${additionalClasses}`} aria-haspopup="true">
        <Button
          buttonType="button"
          className={promptClassName}
          icon={icon}
          id={id}
          isIconOnly={isIconOnly}
          label={label}
          onBlur={() => this.toggleTout()}
          onClick={onClick || (() => this.toggleTout())}
        />
        <div role="tooltip" aria-labelledby={id} aria-hidden="true" className={popupName}>
          {children}
        </div>
      </div>
    );
  }
}

Tout.propTypes = {
  additionalClasses: PropTypes.string,
  alertType: PropTypes.oneOf(['info', 'error', 'alert', 'confirmation']),
  children: PropTypes.node,
  icon: PropTypes.string,
  id: PropTypes.string.isRequired,
  isIconOnly: PropTypes.bool,
  isInline: PropTypes.bool,
  isToolTip: PropTypes.bool,
  label: PropTypes.string.isRequired,
  labelClassName: PropTypes.string,
  onClick: PropTypes.func,
  popupClassName: PropTypes.string,
  position: PropTypes.oneOf([
    'topLeft',
    'topCenter',
    'topRight',
    'bottomLeft',
    'bottomCenter',
    'bottomRight',
    'rightCenter',
    'leftCenter',
  ]),
  size: PropTypes.string,
};

Tout.defaultProps = {
  additionalClasses: '',
  alertType: 'info',
  children: null,
  icon: '',
  isIconOnly: false,
  isInline: false,
  isToolTip: true,
  labelClassName: '',
  onClick: null,
  popupClassName: '',
  position: 'topCenter',
  size: '',
};

const ToutWithClickOutside = onClickOutside(Tout);

export { ToutWithClickOutside as Tout };
