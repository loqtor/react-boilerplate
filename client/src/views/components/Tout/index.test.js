import React from 'react';
import { mount } from 'enzyme';
import axe from 'axe-core';
import toJson from 'enzyme-to-json';

import mountToDoc from '../../../tools/utilities/mountToDoc';

import Tout from './index';

const setup = (render, props) => {
  const defaultProps = {
    label: 'Hello',
    id: 'Tout-helloWorld',
    children: <p>World</p>,
  };

  const component = render(<Tout {...defaultProps} {...props} />);
  const wrappedComponent = render(
    <div className="Tout-wrapper">
      <Tout {...defaultProps} {...props} />
    </div>,
  );

  return {
    actual: component,
    wrappedActual: wrappedComponent,
  };
};

describe('<Tout />', () => {
  it('renders correctly', () => {
    const { actual } = setup(mount, {});

    expect(toJson(actual)).toMatchSnapshot();
  });

  it('renders correctly when small variety', () => {
    const { actual } = setup(mount, { size: 'small' });

    expect(toJson(actual)).toMatchSnapshot();
  });

  it('renders correctly when inline variety', () => {
    const { actual } = setup(mount, { isInline: true });

    expect(toJson(actual)).toMatchSnapshot();
  });

  it('applies the *is-open* class & opens on button click', () => {
    const { actual } = setup(mount, { children: null });

    actual.find('button').simulate('click');

    expect(toJson(actual)).toMatchSnapshot();
    expect(actual.find('.Tout').hasClass('is-open')).toBeTruthy();
  });

  it('removes the *is-open* class & closes on click outside', () => {
    const { wrappedActual, actual } = setup(mount, { children: null });

    wrappedActual.find('.Tout-wrapper').simulate('click');

    expect(toJson(actual)).toMatchSnapshot();
    expect(actual.find('.Tout').hasClass('is-open')).toBeFalsy();
  });

  // it('has no accessibility violations', (done) => {
  //   const { actual } = setup(mountToDoc, {});
  //   const componentNode = actual.getDOMNode();
  //
  //   axe.run(componentNode, (err, { violations }) => {
  //     expect(err).toBe(null);
  //     expect(violations).toHaveLength(0);
  //     done();
  //   });
  // });
});
