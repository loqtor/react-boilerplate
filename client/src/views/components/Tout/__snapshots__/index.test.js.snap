// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tout /> applies the *is-open* class & opens on button click 1`] = `
<OnClickOutside(Tout)
  eventTypes={
    Array [
      "mousedown",
      "touchstart",
    ]
  }
  excludeScrollbar={false}
  id="Tout-helloWorld"
  label="Hello"
  outsideClickIgnoreClass="ignore-react-onclickoutside"
  preventDefault={false}
  stopPropagation={false}
>
  <Tout
    additionalClasses=""
    alertType="info"
    disableOnClickOutside={[Function]}
    enableOnClickOutside={[Function]}
    eventTypes={
      Array [
        "mousedown",
        "touchstart",
      ]
    }
    icon=""
    id="Tout-helloWorld"
    isIconOnly={false}
    isInline={false}
    isToolTip={true}
    label="Hello"
    labelClassName=""
    onClick={null}
    outsideClickIgnoreClass="ignore-react-onclickoutside"
    popupClassName=""
    position="topCenter"
    preventDefault={false}
    size=""
    stopPropagation={false}
  >
    <div
      className="Tout is-open Tout--info Tout--toolTip"
      id="Tout-helloWorld"
    >
      <Button
        buttonType="button"
        className="Tout-prompt"
        disabledClass="Button is-disabled"
        icon=""
        iconClass={null}
        id={null}
        isDisabled={false}
        isFauxDisabled={false}
        isIconOnly={false}
        isLinkButton={false}
        isLoading={false}
        isRoute={false}
        label="Hello"
        onBlur={[Function]}
        onClick={[Function]}
        role={null}
        shouldOpennewTab={false}
        title=""
        url=""
        wrapperClass=""
      >
        <button
          className="Tout-prompt"
          disabled={false}
          id={null}
          onClick={[Function]}
          role={null}
          title=""
          type="button"
        >
          <span
            className=""
          >
            Hello
          </span>
        </button>
      </Button>
      <div
        aria-hidden="true"
        className="Tout-popup is-topCenter"
        role="tooltip"
      >
        Hello
      </div>
    </div>
  </Tout>
</OnClickOutside(Tout)>
`;

exports[`<Tout /> removes the *is-open* class & closes on click outside 1`] = `
<OnClickOutside(Tout)
  eventTypes={
    Array [
      "mousedown",
      "touchstart",
    ]
  }
  excludeScrollbar={false}
  id="Tout-helloWorld"
  label="Hello"
  outsideClickIgnoreClass="ignore-react-onclickoutside"
  preventDefault={false}
  stopPropagation={false}
>
  <Tout
    additionalClasses=""
    alertType="info"
    disableOnClickOutside={[Function]}
    enableOnClickOutside={[Function]}
    eventTypes={
      Array [
        "mousedown",
        "touchstart",
      ]
    }
    icon=""
    id="Tout-helloWorld"
    isIconOnly={false}
    isInline={false}
    isToolTip={true}
    label="Hello"
    labelClassName=""
    onClick={null}
    outsideClickIgnoreClass="ignore-react-onclickoutside"
    popupClassName=""
    position="topCenter"
    preventDefault={false}
    size=""
    stopPropagation={false}
  >
    <div
      className="Tout Tout--info Tout--toolTip"
      id="Tout-helloWorld"
    >
      <Button
        buttonType="button"
        className="Tout-prompt"
        disabledClass="Button is-disabled"
        icon=""
        iconClass={null}
        id={null}
        isDisabled={false}
        isFauxDisabled={false}
        isIconOnly={false}
        isLinkButton={false}
        isLoading={false}
        isRoute={false}
        label="Hello"
        onBlur={[Function]}
        onClick={[Function]}
        role={null}
        shouldOpennewTab={false}
        title=""
        url=""
        wrapperClass=""
      >
        <button
          className="Tout-prompt"
          disabled={false}
          id={null}
          onClick={[Function]}
          role={null}
          title=""
          type="button"
        >
          <span
            className=""
          >
            Hello
          </span>
        </button>
      </Button>
      <div
        aria-hidden="true"
        className="Tout-popup is-topCenter"
        role="tooltip"
      >
        Hello
      </div>
    </div>
  </Tout>
</OnClickOutside(Tout)>
`;

exports[`<Tout /> renders correctly 1`] = `
<OnClickOutside(Tout)
  eventTypes={
    Array [
      "mousedown",
      "touchstart",
    ]
  }
  excludeScrollbar={false}
  id="Tout-helloWorld"
  label="Hello"
  outsideClickIgnoreClass="ignore-react-onclickoutside"
  preventDefault={false}
  stopPropagation={false}
>
  <Tout
    additionalClasses=""
    alertType="info"
    disableOnClickOutside={[Function]}
    enableOnClickOutside={[Function]}
    eventTypes={
      Array [
        "mousedown",
        "touchstart",
      ]
    }
    icon=""
    id="Tout-helloWorld"
    isIconOnly={false}
    isInline={false}
    isToolTip={true}
    label="Hello"
    labelClassName=""
    onClick={null}
    outsideClickIgnoreClass="ignore-react-onclickoutside"
    popupClassName=""
    position="topCenter"
    preventDefault={false}
    size=""
    stopPropagation={false}
  >
    <div
      className="Tout Tout--info Tout--toolTip"
      id="Tout-helloWorld"
    >
      <p>
        World
      </p>
      <div
        aria-hidden="true"
        className="Tout-popup is-topCenter"
        role="tooltip"
      >
        Hello
      </div>
    </div>
  </Tout>
</OnClickOutside(Tout)>
`;

exports[`<Tout /> renders correctly when inline variety 1`] = `
<OnClickOutside(Tout)
  eventTypes={
    Array [
      "mousedown",
      "touchstart",
    ]
  }
  excludeScrollbar={false}
  id="Tout-helloWorld"
  isInline={true}
  label="Hello"
  outsideClickIgnoreClass="ignore-react-onclickoutside"
  preventDefault={false}
  stopPropagation={false}
>
  <Tout
    additionalClasses=""
    alertType="info"
    disableOnClickOutside={[Function]}
    enableOnClickOutside={[Function]}
    eventTypes={
      Array [
        "mousedown",
        "touchstart",
      ]
    }
    icon=""
    id="Tout-helloWorld"
    isIconOnly={false}
    isInline={true}
    isToolTip={true}
    label="Hello"
    labelClassName=""
    onClick={null}
    outsideClickIgnoreClass="ignore-react-onclickoutside"
    popupClassName=""
    position="topCenter"
    preventDefault={false}
    size=""
    stopPropagation={false}
  >
    <div
      className="Tout Tout--info Tout--inline Tout--toolTip"
      id="Tout-helloWorld"
      role="tooltip"
    >
      <span
        className="icon icon-info"
      />
      <p>
        World
      </p>
    </div>
  </Tout>
</OnClickOutside(Tout)>
`;

exports[`<Tout /> renders correctly when small variety 1`] = `
<OnClickOutside(Tout)
  eventTypes={
    Array [
      "mousedown",
      "touchstart",
    ]
  }
  excludeScrollbar={false}
  id="Tout-helloWorld"
  label="Hello"
  outsideClickIgnoreClass="ignore-react-onclickoutside"
  preventDefault={false}
  size="small"
  stopPropagation={false}
>
  <Tout
    additionalClasses=""
    alertType="info"
    disableOnClickOutside={[Function]}
    enableOnClickOutside={[Function]}
    eventTypes={
      Array [
        "mousedown",
        "touchstart",
      ]
    }
    icon=""
    id="Tout-helloWorld"
    isIconOnly={false}
    isInline={false}
    isToolTip={true}
    label="Hello"
    labelClassName=""
    onClick={null}
    outsideClickIgnoreClass="ignore-react-onclickoutside"
    popupClassName=""
    position="topCenter"
    preventDefault={false}
    size="small"
    stopPropagation={false}
  >
    <div
      className="Tout Tout--info Tout--toolTip"
      id="Tout-helloWorld"
    >
      <p>
        World
      </p>
      <div
        aria-hidden="true"
        className="Tout-popup is-topCenter"
        role="tooltip"
      >
        Hello
      </div>
    </div>
  </Tout>
</OnClickOutside(Tout)>
`;
